glib-macros/src/shared_derive.rs

```diff
@@ -0,0 +1,191 @@
+            }
+        }
+    }
+}
+
+fn gen_ptr_to_option(name: &Ident, inner_name: &syn::Path, nullable: bool) -> TokenStream {
+    if nullable {
+        quote! {
+            if ptr.is_null() {
+                None
+            } else {
+                Some(#name(std::sync::Arc::from_raw(ptr as *const #inner_name)))
+            }
+        }
+    } else {
+        quote! {
+            assert!(!ptr.is_null());
+            Some(#name(std::sync::Arc::from_raw(ptr as *const #inner_name)))
+        }
+    }
+}
+
+fn inner_name(input: &syn::DeriveInput) -> Option<&syn::Path> {
+    let fields = match &input.data {
+        syn::Data::Struct(s) => &s.fields,
+        _ => return None,
+    };
+
+    let unnamed = match fields {
+        syn::Fields::Unnamed(u) if u.unnamed.len() == 1 => &u.unnamed[0],
+        _ => return None,
+    };
+
+    let path = match &unnamed.ty {
+        syn::Type::Path(p) => p,
+        _ => return None,
+    };
+
+    let inner = match path.path.segments.last() {
+        Some(s) if s.ident == "Arc" => &s.arguments,
+        _ => return None,
+    };
+
+    let inner = match &inner {
+        syn::PathArguments::AngleBracketed(a) if a.args.len() == 1 => &a.args[0],
+        _ => return None,
+    };
+
+    let inner = match &inner {
+        syn::GenericArgument::Type(syn::Type::Path(p)) => &p.path,
+        _ => return None,
+    };
+
+    Some(inner)
+}
+
+pub fn impl_shared(input: &syn::DeriveInput) -> proc_macro2::TokenStream {
+    let inner_name = match inner_name(input) {
+        Some(p) => p,
+        _ => abort_call_site!("derive(Shared) requires struct MyStruct(Arc<MyInnerStruct>)"),
+    };
+
+    let name = &input.ident;
+    let gtype_name = match parse_type_name(&input, "shared") {
+        Ok(v) => v,
+        Err(e) => abort_call_site!(
+            "{}: derive(Shared) requires #[shared(type_name = \"SharedTypeName\")]",
+            e
+        ),
+    };
+
+    let meta = find_attribute_meta(&input.attrs, "shared")
+        .unwrap()
+        .unwrap();
+    let nullable = find_nested_meta(&meta, "nullable").is_some();
+    let crate_ident = crate_ident_new();
+    let ptr_to_option = gen_ptr_to_option(name, inner_name, nullable);
+
+    let impl_from_value = if !nullable {
+        gen_impl_from_value(name, inner_name, &crate_ident)
+    } else {
+        quote! {}
+    };
+
+    let impl_set_value_optional = if nullable {
+        gen_impl_set_value_optional(name, inner_name, &crate_ident)
+    } else {
+        quote! {}
+    };
+
+    quote! {
+        impl #crate_ident::subclass::shared::SharedType for #name {
+            const NAME: &'static str = #gtype_name;
+
+            type InnerType = #inner_name;
+
+            fn ref_(this: *const Self::InnerType) -> *const Self::InnerType {
+                unsafe {
+                    use std::mem::ManuallyDrop;
+                    let this_arc = ManuallyDrop::new(std::sync::Arc::from_raw(this));
```

Comment:
There's `Arc::decrement_strong_count` (and increment) that directly works on the pointer. That should be easier :)

Reply:
That requires rust 1.51 though... do we want to bump the min rust version?

Reply:
I don't mind as many of my projects and/or their dependencies require that anyway, but maybe @GuillaumeGomez has an opinion here? :) tl;rd: 1.51 would allow to do things cleaner here, older versions require workarounds.

Reply:
Let upgrade the version then. If that allows to remove dependencies, let's go for it!

Reply:
It does not remove dependencies though.
I am personally ok with requiring new rustc if we have good use cases for const generics etc, but bumping to save one line here seems excessive tbh :)

Reply:
I had more places before where I would've liked to use newer APIs but didn't for the same reason :) At some point it makes sense to just update.